library/std/src/sys/windows/thread_local_key.rs

```diff
@@ -19,106 +21,93 @@ pub type Dtor = unsafe extern "C" fn(*mut u8);
 // somewhere to run arbitrary code on thread termination. With this in place
 // we'll be able to run anything we like, including all TLS destructors!
 //
-// To accomplish this feat, we perform a number of threads, all contained
-// within this module:
-//
-// * All TLS destructors are tracked by *us*, not the windows runtime. This
-//   means that we have a global list of destructors for each TLS key that
-//   we know about.
-// * When a thread exits, we run over the entire list and run dtors for all
-//   non-null keys. This attempts to match Unix semantics in this regard.
-//
-// This ends up having the overhead of using a global list, having some
-// locks here and there, and in general just adding some more code bloat. We
-// attempt to optimize runtime by forgetting keys that don't have
-// destructors, but this only gets us so far.
+// Since the maximum number of keys is 1088 [3] and key values are always lower
+// than 1088 [4], we can just use a static array to store the destructor functions
+// and use the TLS key as index. This avoids all synchronization problems
+// encountered with linked lists or other kinds of storage.
 //
 // For more details and nitty-gritty, see the code sections below!
 //
 // [1]: https://www.codeproject.com/Articles/8113/Thread-Local-Storage-The-C-Way
-// [2]: https://github.com/ChromiumWebApps/chromium/blob/master/base
-//                        /threading/thread_local_storage_win.cc#L42
+// [2]: https://github.com/ChromiumWebApps/chromium/blob/master/base/threading/thread_local_storage_win.cc#L42
+// [3]: https://learn.microsoft.com/en-us/windows/win32/procthread/thread-local-storage
+// [4]: https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-tlssetvalue
 
-// -------------------------------------------------------------------------
-// Native bindings
-//
-// This section is just raw bindings to the native functions that Windows
-// provides, There's a few extra calls to deal with destructors.
+static DTORS: [AtomicPtr<()>; 1088] = [const { AtomicPtr::new(ptr::null_mut()) }; 1088];
+// The highest key that has a destructor associated with it. Used as an
+// optimization so we don't need to iterate over the whole array when there
+// are only a few keys.
+static HIGHEST: AtomicUsize = AtomicUsize::new(0);
 
 #[inline]
 pub unsafe fn create(dtor: Option<Dtor>) -> Key {
     let key = c::TlsAlloc();
     assert!(key != c::TLS_OUT_OF_INDEXES);
-    if let Some(f) = dtor {
-        register_dtor(key, f);
+
+    if let Some(dtor) = dtor {
+        DTORS[key as usize].store(mem::transmute::<Dtor, *mut ()>(dtor), Relaxed);
+        HIGHEST.fetch_max(key as usize, Relaxed);
+        // If the destructors are run in a signal handler running after this
+        // code, we need to guarantee that the changes have been performed
+        // before the handler is triggered.
+        compiler_fence(Release);
```

Comment:
This micro-optimization doesn't really feel likely to be worth it, and it's plausibly not even totally sound.

We should just use the correct orderings for these atomics -- it really shouldn't make a performance difference and is easier to reason about.